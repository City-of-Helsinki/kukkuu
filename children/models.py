from hashlib import sha256
from typing import Iterable, Optional

from django.conf import settings
from django.core.validators import RegexValidator
from django.db import models, transaction
from django.db.models import Q
from django.utils import timezone
from django.utils.translation import gettext_lazy as _

from common.models import TimestampedModel, UUIDPrimaryKeyModel
from languages.models import Language
from users.models import Guardian


class ChildQuerySet(models.QuerySet):
    def user_can_view(self, user):
        return self.filter(
            Q(guardians__user=user) | Q(project__in=user.administered_projects)
        ).distinct()

    def user_can_update(self, user):
        return self.filter(guardians__user=user)

    def user_can_delete(self, user):
        return self.filter(guardians__user=user)

    @transaction.atomic()
    def delete(self):
        for child in self:
            child.delete()


postal_code_validator = RegexValidator(
    regex=r"^\d{5}$", message="Postal code must be 5 digits", code="invalid_postal_code"
)


class Child(UUIDPrimaryKeyModel, TimestampedModel):
    name = models.CharField(verbose_name=_("name"), max_length=64, blank=True)
    birthyear = models.DateField(verbose_name=_("birthyear"))
    postal_code = models.CharField(
        verbose_name=_("postal code"),
        max_length=5,
        validators=[postal_code_validator],
    )
    guardians = models.ManyToManyField(
        Guardian,
        verbose_name=_("guardians"),
        related_name="children",
        through="children.Relationship",
        blank=True,
    )
    project = models.ForeignKey(
        "projects.Project",
        verbose_name=_("project"),
        related_name="children",
        on_delete=models.PROTECT,
    )
    languages_spoken_at_home = models.ManyToManyField(
        Language,
        verbose_name=_("languages spoken at home"),
        related_name="children",
        blank=True,
    )

    objects = ChildQuerySet.as_manager()

    class Meta:
        verbose_name = _("child")
        verbose_name_plural = _("children")
        ordering = ["birthyear", "name"]

    def __str__(self):
        return f"{self.name} ({self.birthyear})"

    @transaction.atomic()
    def delete(self, *args, **kwargs):
        self.enrolments.upcoming().delete()
        return super().delete(*args, **kwargs)

    def can_user_administer(self, user):
        return user.can_administer_project(self.project)

    def get_enrolment_count(self, year: Optional[int] = None, past=False):
        from events.models import Event  # noqa

        if year and past:
            raise ValueError("Cannot use year and past arguments at the same time.")
        now = timezone.now()
        year = year or now.year

        occurrence_filters = Q(time__year=year)
        # For external ticket system events published_at field is used to determine the
        # event's year. That is obviously not a perfect solution, but the best we can
        # do with the current data, and most probably good enough.
        ticket_system_filters = Q(event__published_at__year=year)
        if past:
            occurrence_filters &= Q(time__lt=now)
            ticket_system_filters &= Q(event__published_at__lt=now)
        return (
            self.occurrences.filter(occurrence_filters).count()
            + self.ticket_system_passwords.filter(ticket_system_filters)
            .distinct()
            .count()
        )

    @staticmethod
    def _hash(values: Iterable[str]) -> str:
        return sha256(",".join(values).encode("utf8")).hexdigest()

    @property
    def _salt_birthyear_postal_code_guardian_emails_list(self) -> list[str]:
        """
        List of salt, child's birthyear, postal code and guardians' emails.
        """
        return [
            settings.KUKKUU_HASHID_SALT,
            self.birthyear.isoformat(),
            self.postal_code,
            *sorted(guardian.email for guardian in self.guardians.all()),
        ]

    @property
    def birthyear_postal_code_guardian_emails_hash(self) -> str:
        """
        Salted hash of child's birthyear, postal code and guardians' emails.
        """
        return self._hash(self._salt_birthyear_postal_code_guardian_emails_list)

    @property
    def name_birthyear_postal_code_guardian_emails_hash(self) -> str:
        """
        Salted hash of child's name, birthyear, postal
        code and guardians' emails.
        """
        return self._hash(
            [self.name] + self._salt_birthyear_postal_code_guardian_emails_list
        )


class RelationshipQuerySet(models.QuerySet):
    def user_can_view(self, user):
        return self.filter(
            Q(guardian__user=user) | Q(child__project__in=user.administered_projects)
        ).distinct()


class Relationship(models.Model):
    PARENT = "parent"  # In Finnish: Vanhempi
    OTHER_GUARDIAN = "other_guardian"  # In Finnish: Muu huoltaja
    OTHER_RELATION = "other_relation"  # In Finnish: Muu l√§heinen
    ADVOCATE = "advocate"  # In Finnish: Virallinen puolesta-asioija
    TYPE_CHOICES = (
        (PARENT, _("Parent")),
        (OTHER_GUARDIAN, _("Other guardian")),
        (OTHER_RELATION, _("Other relation")),
        (ADVOCATE, _("Advocate")),
    )

    created_at = models.DateTimeField(verbose_name=_("created at"), auto_now_add=True)
    child = models.ForeignKey(
        Child,
        verbose_name=_("child"),
        related_name="relationships",
        on_delete=models.CASCADE,
    )
    guardian = models.ForeignKey(
        Guardian,
        verbose_name=_("guardian"),
        on_delete=models.CASCADE,
        related_name="relationships",
    )
    type = models.CharField(
        verbose_name=_("type"),
        choices=TYPE_CHOICES,
        max_length=64,
        null=True,
        blank=True,
    )

    objects = RelationshipQuerySet.as_manager()

    class Meta:
        verbose_name = _("relationship")
        verbose_name_plural = _("relationships")
        ordering = [
            "guardian",
            "child",
        ]
