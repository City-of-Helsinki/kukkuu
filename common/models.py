import uuid

from django.db import models
from django.utils.translation import ugettext_lazy as _
from parler.managers import TranslatableQuerySet as ParlerTranslatableQuerySet
from parler.models import TranslatableModel as ParlerTranslatableModel


class TimestampedModel(models.Model):
    created_at = models.DateTimeField(verbose_name=_("created at"), auto_now_add=True)
    updated_at = models.DateTimeField(verbose_name=_("updated_at"), auto_now=True)

    class Meta:
        abstract = True


class UUIDPrimaryKeyModel(models.Model):
    id = models.UUIDField(
        verbose_name=_("UUID"), primary_key=True, default=uuid.uuid4, editable=False
    )

    class Meta:
        abstract = True


class TranslatableQuerySet(ParlerTranslatableQuerySet):
    def create_translatable_object(self, **kwargs):
        translations = kwargs.pop("translations")
        obj = self.create(**kwargs)
        for translation in translations:
            language_code = translation.pop("language_code")
            obj.create_translation(language_code=language_code, **translation)
        return obj


class TranslatableModel(ParlerTranslatableModel):
    objects = TranslatableQuerySet.as_manager()

    class Meta:
        abstract = True
