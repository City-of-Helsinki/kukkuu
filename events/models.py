from django.conf import settings
from django.db import models
from django.db.models import Q
from django.utils import timezone
from django.utils.timezone import now
from django.utils.translation import ugettext_lazy as _
from parler.models import TranslatedFields

from children.models import Child
from common.models import TimestampedModel, TranslatableModel, TranslatableQuerySet
from events.consts import NotificationType
from events.utils import send_event_notifications_to_guardians
from venues.models import Venue


# This need to be inherited from TranslatableQuerySet instead of default model.QuerySet
class EventQueryset(TranslatableQuerySet):
    def user_can_view(self, user):
        return self.filter(
            Q(project__users=user) | Q(published_at__isnull=False)
        ).distinct()

    def published(self):
        return self.filter(published_at__isnull=False)


class Event(TimestampedModel, TranslatableModel):
    CHILD_AND_GUARDIAN = "child_and_guardian"
    CHILD_AND_1_OR_2_GUARDIANS = "child_and_1_or_2_guardians"
    FAMILY = "family"
    PARTICIPANTS_PER_INVITE_CHOICES = (
        (CHILD_AND_GUARDIAN, _("Child and guardian")),
        (CHILD_AND_1_OR_2_GUARDIANS, _("Child and 1-2 guardians")),
        (FAMILY, _("Family")),
    )

    translations = TranslatedFields(
        name=models.CharField(verbose_name=_("name"), max_length=255, blank=True),
        short_description=models.TextField(
            verbose_name=_("short description"), blank=True
        ),
        description=models.TextField(verbose_name=_("description"), blank=True),
        image_alt_text=models.CharField(
            verbose_name=_("image alt text"), blank=True, max_length=255
        ),
    )
    image = models.ImageField(blank=True, verbose_name=_("image"))
    participants_per_invite = models.CharField(
        max_length=255,
        choices=PARTICIPANTS_PER_INVITE_CHOICES,
        verbose_name=_("participants per invite"),
    )
    duration = models.PositiveSmallIntegerField(
        verbose_name=_("duration"), blank=True, null=True, help_text=_("In minutes")
    )
    capacity_per_occurrence = models.PositiveSmallIntegerField(
        verbose_name=_("capacity per occurrence")
    )
    published_at = models.DateTimeField(
        blank=True, null=True, verbose_name=_("published at")
    )

    project = models.ForeignKey(
        "projects.Project",
        verbose_name=_("project"),
        related_name="events",
        on_delete=models.CASCADE,
    )

    objects = EventQueryset.as_manager()

    class Meta:
        verbose_name = _("event")
        verbose_name_plural = _("events")

    def __str__(self):
        name = self.safe_translation_getter("name", super().__str__())
        published_text = _("published") if self.published_at else _("unpublished")
        return f"{name} ({self.pk}) ({self.project.year}) ({published_text})"

    def can_user_administer(self, user):
        return user.projects.filter(pk=self.project_id).exists()

    def publish(self):
        self.published_at = timezone.now()
        self.save()

        send_event_notifications_to_guardians(
            self,
            NotificationType.EVENT_PUBLISHED,
            self.project.children.prefetch_related("guardians"),
        )

    def is_published(self):
        return bool(self.published_at)


class OccurrenceQueryset(models.QuerySet):
    def user_can_view(self, user):
        return self.filter(
            Q(event__project__users=user) | Q(event__published_at__isnull=False)
        ).distinct()

    def delete(self, *args, **kwargs):
        for obj in self:
            obj.delete()


class Occurrence(TimestampedModel):
    time = models.DateTimeField(verbose_name=_("time"))
    event = models.ForeignKey(
        Event,
        verbose_name=_("event"),
        related_name="occurrences",
        on_delete=models.CASCADE,
    )
    venue = models.ForeignKey(
        Venue,
        verbose_name=_("venue"),
        related_name="occurrences",
        on_delete=models.CASCADE,
    )

    children = models.ManyToManyField(
        Child,
        verbose_name=_("children"),
        related_name="occurrences",
        through="events.Enrolment",
        blank=True,
    )
    occurrence_language = models.CharField(
        max_length=10,
        choices=settings.LANGUAGES,
        verbose_name=_("occurrence language"),
        default=settings.LANGUAGES[0][0],
    )
    objects = OccurrenceQueryset.as_manager()

    class Meta:
        verbose_name = _("occurrence")
        verbose_name_plural = _("occurrences")

    def __str__(self):
        return f"{self.time} ({self.pk})"

    def delete(self, *args, **kwargs):
        if self.time >= now():
            # this QS needs to be evaluated here, it would not work after the
            # occurrence has been deleted
            children = list(self.children.all())

            super().delete(*args, **kwargs)

            send_event_notifications_to_guardians(
                self.event,
                NotificationType.OCCURRENCE_CANCELLED,
                children,
                occurrence=self,
            )
        else:
            super().delete(*args, **kwargs)

    def get_enrolment_count(self):
        try:
            # try to use an annotated value
            return self.enrolment_count
        except AttributeError:
            return self.enrolments.count()

    def can_user_administer(self, user):
        # There shouldn't ever be a situation where event.project != venue.project
        # so we can just check one of them
        return user.projects.filter(pk=self.event.project.pk).exists()


class Enrolment(TimestampedModel):
    child = models.ForeignKey(
        Child,
        related_name="enrolments",
        on_delete=models.CASCADE,
        verbose_name=_("child"),
    )
    occurrence = models.ForeignKey(
        Occurrence,
        related_name="enrolments",
        on_delete=models.CASCADE,
        verbose_name=_("occurrence"),
    )
    attended = models.NullBooleanField(verbose_name=_("attended"))

    class Meta:
        verbose_name = _("enrolment")
        verbose_name_plural = _("enrolments")
        constraints = [
            models.UniqueConstraint(
                fields=["child", "occurrence"], name="unq_child_occurrence"
            )
        ]

    def __str__(self):
        return f"{self.pk} {self.child_id}"

    def save(self, *args, **kwargs):
        created = self.pk is None
        super().save(*args, **kwargs)

        if created:
            send_event_notifications_to_guardians(
                self.occurrence.event,
                NotificationType.OCCURRENCE_ENROLMENT,
                self.child,
                occurrence=self.occurrence,
            )

    def delete_and_send_notification(self):
        child = self.child
        occurrence = self.occurrence
        self.delete()

        send_event_notifications_to_guardians(
            occurrence.event,
            NotificationType.OCCURRENCE_UNENROLMENT,
            child,
            occurrence=occurrence,
        )

    def can_user_administer(self, user):
        return self.occurrence.can_user_administer(user)
