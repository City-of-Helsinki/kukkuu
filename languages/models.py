import gettext

import pycountry
from django.core.exceptions import ValidationError
from django.db import IntegrityError, models, transaction
from django.utils import translation
from django.utils.translation import ugettext_lazy as _
from parler.models import TranslatableModel, TranslatedFields

from common.models import TranslatableQuerySet
from kukkuu.settings import PARLER_SUPPORTED_LANGUAGE_CODES


class InvalidLanguageCodeError(Exception):
    pass


class LanguageAlreadyExistsError(Exception):
    pass


class LanguageQueryset(TranslatableQuerySet):
    @transaction.atomic
    def create_from_language_code(self, language_code):
        pycountry_language = get_pycountry_language(language_code)

        with translation.override("en"):
            try:
                obj = self.create(
                    alpha_3_code=pycountry_language.alpha_3,
                    name=pycountry_language.name,
                )
            except IntegrityError as e:
                raise LanguageAlreadyExistsError(e)

        for parler_language in PARLER_SUPPORTED_LANGUAGE_CODES:
            if parler_language == "en":
                continue

            gettext_translation = gettext.translation(
                "iso639-3", pycountry.LOCALES_DIR, languages=[parler_language]
            )
            obj.create_translation(
                parler_language,
                name=gettext_translation.gettext(pycountry_language.name),
            )

        return obj

    @transaction.atomic
    def create_option_other(self):
        obj, created = self.language("en").get_or_create(
            alpha_3_code=None,
            translations__name="Other language",
            defaults={"name": "Other language"},
        )
        if not created:
            raise LanguageAlreadyExistsError()

        obj.create_translation("sv", name="Annat spr√•k")
        obj.create_translation("fi", name="muu kieli")

        return obj


class Language(TranslatableModel):
    alpha_3_code = models.CharField(
        max_length=3,
        verbose_name=_("alpha-3 code"),
        help_text=_("ISO 639-3 (language) or ISO 639-5 (language family) alpha-3 code"),
        unique=True,
        null=True,
        blank=True,
    )
    translations = TranslatedFields(
        name=models.CharField(verbose_name=_("name"), max_length=255, blank=True),
    )

    objects = LanguageQueryset.as_manager()

    class Meta:
        verbose_name = _("language")
        verbose_name_plural = _("languages")

    def __str__(self):
        return f"{self.name} ({self.alpha_3_code})"

    def clean(self):
        if self.alpha_3_code:
            try:
                get_pycountry_language(self.alpha_3_code)
            except InvalidLanguageCodeError as e:
                raise ValidationError({"alpha_3_code": e})


def get_pycountry_language(language_code):
    language = pycountry.languages.get(
        alpha_3=language_code
    ) or pycountry.language_families.get(alpha_3=language_code)

    if not language:
        raise InvalidLanguageCodeError(f'Invalid language code "{language_code}"')

    return language
