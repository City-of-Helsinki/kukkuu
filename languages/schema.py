import graphene
from django.apps import apps
from graphene import relay
from graphene_django import DjangoConnectionField, DjangoObjectType
from languages.models import Language

LanguageTranslation = apps.get_model("languages", "LanguageTranslation")


class LanguageTranslationType(DjangoObjectType):
    class Meta:
        model = LanguageTranslation
        exclude = ("id", "master")


class LanguageNode(DjangoObjectType):
    name = graphene.String()

    class Meta:
        model = Language
        interfaces = (relay.Node,)
        fields = ("alpha_3_code", "name", "translations")

    @classmethod
    def get_queryset(cls, queryset, info):
        return queryset.prefetch_related("translations")


class Query:
    language = relay.Node.Field(LanguageNode)
    languages = DjangoConnectionField(LanguageNode)
