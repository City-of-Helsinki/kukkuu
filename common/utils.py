import binascii
from copy import deepcopy
from functools import wraps

from django.conf import settings
from django.core.exceptions import PermissionDenied
from django.db import transaction
from graphene import Node
from graphql.execution.base import ResolveInfo
from graphql_relay import from_global_id, to_global_id

from kukkuu import __version__
from kukkuu.exceptions import DataValidationError, ObjectDoesNotExistError
from kukkuu.settings import REVISION


def update_object(obj, data):
    if not data:
        return
    for k, v in data.items():
        if v is None and not obj.__class__._meta.get_field(k).null:
            raise DataValidationError(f"{k} cannot be null.")
        setattr(obj, k, v)
    obj.save()


@transaction.atomic
def update_object_with_translations(model, model_data):
    model_data = deepcopy(model_data)
    translations_input = model_data.pop("translations", None)
    if translations_input:
        model.create_or_update_translations(translations_input)
    update_object(model, model_data)


def get_api_version():
    return " | ".join((__version__, REVISION))


def get_global_id(obj):
    return to_global_id(f"{obj.__class__.__name__}Node", obj.pk)


def get_node_id_from_global_id(global_id, expected_node_name):
    if not global_id:
        return None
    try:
        name, id = from_global_id(global_id)
    except (
        binascii.Error,
        UnicodeDecodeError,
    ):  # invalid global ID
        return None
    return id if name == expected_node_name else None


def check_can_user_administer(obj, user):
    try:
        yes_we_can = obj.can_user_administer(user)
    except AttributeError:
        raise TypeError(
            f"{obj.__class__.__name__} model does not implement can_user_administer()."
        )
    if not yes_we_can:
        raise PermissionDenied()


def get_obj_from_global_id(info, global_id, expected_obj_type):
    obj = Node.get_node_from_global_id(info, global_id)
    if not obj or type(obj) is not expected_obj_type:
        raise ObjectDoesNotExistError(
            f"{expected_obj_type.__name__} with ID {global_id} does not exist."
        )
    return obj


def get_obj_if_user_can_administer(info, global_id, expected_obj_type):
    obj = get_obj_from_global_id(info, global_id, expected_obj_type)
    check_can_user_administer(obj, info.context.user)
    return obj


def context(f):
    def decorator(func):
        def wrapper(*args, **kwargs):
            info = next(arg for arg in args if isinstance(arg, ResolveInfo))
            return func(info.context, *args, **kwargs)

        return wrapper

    return decorator


def user_passes_test(test_func):
    def decorator(f):
        @wraps(f)
        @context(f)
        def wrapper(context, *args, **kwargs):
            if test_func(context.user):
                return f(*args, **kwargs)
            raise PermissionDenied("You do not have permission to perform this action")

        return wrapper

    return decorator


# copied from https://github.com/flavors/django-graphql-jwt/blob/704f24e7ebbea0b81015ef3c1f4a302e9d432ecf/graphql_jwt/decorators.py  # noqa
login_required = user_passes_test(lambda u: u.is_authenticated)

project_user_required = user_passes_test(
    lambda u: u.is_authenticated and u.administered_projects
)


def get_translations_dict(obj, field_name):
    """
    Returns a dict of translations for a given field of a model instance.

    :param obj: The model instance
    :param field_name: The name of the field
    :return: A dict with language codes as keys and translations as values
    """

    return {
        lang_code: getattr(
            obj.translations.filter(language_code=lang_code).first(), field_name, ""
        )
        for lang_code, _ in settings.LANGUAGES
    }
