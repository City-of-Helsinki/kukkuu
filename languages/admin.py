from django.contrib import admin
from parler.admin import TranslatableAdmin

from .models import Language


@admin.register(Language)
class LanguageAdmin(TranslatableAdmin):
    list_display = ("alpha_3_code", "name")
