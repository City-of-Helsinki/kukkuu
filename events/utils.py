import logging
from collections.abc import Iterable
from datetime import datetime
from typing import List, Optional, TYPE_CHECKING, Union

from django.conf import settings
from django.db.models import QuerySet
from django.utils import timezone
from django_ilmoitin.utils import send_notification
from parler.utils.context import switch_language

from common.utils import get_global_id
from users.utils import get_communication_unsubscribe_ui_url

if TYPE_CHECKING:
    from children.models import Child
    from events.consts import NotificationType
    from events.models import Event, EventGroup

logger = logging.getLogger(__name__)


def send_event_notifications_to_guardians(
    event: "Event",
    notification_type: "NotificationType",
    children: Union[QuerySet, List["Child"]],
    attachments: Optional[List] = None,
    **kwargs,
):
    if not isinstance(children, Iterable):
        children = [children]

    context_for_address = {}

    for child in children:
        for guardian in child.guardians.all():
            with switch_language(event, guardian.language):
                context = {
                    "event": event,
                    "child": child,
                    "guardian": guardian,
                    "event_url": get_event_ui_url(event, child, guardian.language),
                    "localtime": timezone.template_localtime,
                    "get_global_id": get_global_id,
                    "unsubscribe_url": get_communication_unsubscribe_ui_url(
                        guardian, guardian.language
                    ),
                    "is_obsolete": child.is_obsolete,
                    **kwargs,
                }
                occurrence = kwargs.get("occurrence")
                if occurrence:
                    context["occurrence_url"] = get_occurrence_ui_url(
                        occurrence, child, guardian.language
                    )
                    context["occurrence_enrol_url"] = get_occurrence_enrol_ui_url(
                        occurrence, child, guardian.language
                    )
                context_for_address[guardian.email] = context

    emails = ",".join(context_for_address.keys())
    logger.debug(
        f"There are {len(emails)} total in the notification list. "
        f"Notification will be sent to these addresses: {emails}"
    )
    # NOTE: Instead of sending the notification in the previous loop,
    # lets first collect the mail contexts to a dictionary and
    # send the mails after all the contexts have been handled properly.
    # This way we can prevent flooding the recipients with the mails
    # when some of the contexts cannot be handled because of any reason.
    # For more details, see
    # 1. https://helsinkisolutionoffice.atlassian.net/browse/KK-984,
    # 2. https://helsinkisolutionoffice.atlassian.net/browse/PT-1414.
    for address, context in context_for_address.items():
        logger.debug(f"Sending event notification to {address}.")
        guardian = context["guardian"]
        send_notification(
            email=address,
            notification_type=notification_type,
            context=context,
            language=guardian.language,
            attachments=attachments,
        )


def send_event_group_notifications_to_guardians(
    event_group: "EventGroup",
    notification_type: "NotificationType",
    children: Union[QuerySet, List["Child"]],
    **kwargs,
):
    if not isinstance(children, Iterable):
        children = [children]

    for child in children:
        for guardian in child.guardians.all():
            with switch_language(event_group, guardian.language):
                context = {
                    "event_group": event_group,
                    "child": child,
                    "guardian": guardian,
                    "event_group_url": get_event_group_ui_url(
                        event_group, child, guardian.language
                    ),
                    "localtime": timezone.template_localtime,
                    "get_global_id": get_global_id,
                    "events": [
                        {
                            "obj": event,
                            "event_url": get_event_ui_url(
                                event, child, guardian.language
                            ),
                        }
                        for event in event_group.events.language(guardian.language)
                    ],
                    "unsubscribe_url": get_communication_unsubscribe_ui_url(
                        guardian, guardian.language
                    ),
                    "is_obsolete": child.is_obsolete,
                    **kwargs,
                }

                send_notification(
                    guardian.email,
                    notification_type,
                    context=context,
                    language=guardian.language,
                )


def convert_to_localtime_tz(value):
    dt = datetime.combine(datetime.now().date(), value)
    if timezone.is_naive(value):
        # Auto add local timezone to naive time
        return timezone.make_aware(dt).timetz()
    else:
        return timezone.localtime(dt).timetz()


def get_event_ui_url(event, child, language):
    return "{}/{}/profile/child/{}/event/{}".format(
        settings.KUKKUU_UI_BASE_URL,
        language,
        get_global_id(child),
        get_global_id(event),
    )


def get_event_group_ui_url(event_group, child, language):
    return "{}/{}/profile/child/{}/event-group/{}".format(
        settings.KUKKUU_UI_BASE_URL,
        language,
        get_global_id(child),
        get_global_id(event_group),
    )


def get_occurrence_ui_url(occurrence, child, language):
    return "{}/{}/profile/child/{}/occurrence/{}".format(
        settings.KUKKUU_UI_BASE_URL,
        language,
        get_global_id(child),
        get_global_id(occurrence),
    )


def get_occurrence_enrol_ui_url(occurrence, child, language):
    return "{}/{}/profile/child/{}/event/{}/occurrence/{}/enrol".format(
        settings.KUKKUU_UI_BASE_URL,
        language,
        get_global_id(child),
        get_global_id(occurrence.event),
        get_global_id(occurrence),
    )
