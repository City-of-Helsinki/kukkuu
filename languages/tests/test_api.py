import pytest
from languages.models import Language

from common.utils import get_global_id


@pytest.fixture(autouse=True)
def autouse_db(db):
    pass


@pytest.fixture(autouse=True)
def languages():
    for code in ("fin", "swe", "eng"):
        Language.objects.create_from_language_code(code)


LANGUAGES_QUERY = """
query Languages {
  languages {
    edges {
      node {
        alpha3Code
        name
        translations {
          languageCode
          name
        }
      }
    }
  }
}
"""

LANGUAGE_QUERY = """
query Language($id: ID!) {
  language(id: $id){
    alpha3Code
    name
    translations {
      languageCode
      name
    }
  }
}
"""


def test_languages_query(snapshot, api_client):
    executed = api_client.execute(LANGUAGES_QUERY)

    snapshot.assert_match(executed)


def test_language_query(snapshot, api_client):
    variables = {"id": get_global_id(Language.objects.get(alpha_3_code="fin"))}

    executed = api_client.execute(LANGUAGE_QUERY, variables=variables)

    snapshot.assert_match(executed)
