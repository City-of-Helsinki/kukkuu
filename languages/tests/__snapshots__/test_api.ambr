# serializer version: 1
# name: test_language_query
  dict({
    'data': dict({
      'language': dict({
        'alpha3Code': 'fin',
        'name': 'suomi',
        'translations': list([
          dict({
            'languageCode': 'EN',
            'name': 'Finnish',
          }),
          dict({
            'languageCode': 'FI',
            'name': 'suomi',
          }),
          dict({
            'languageCode': 'SV',
            'name': 'Finska',
          }),
        ]),
      }),
    }),
  })
# ---
# name: test_languages_query
  dict({
    'data': dict({
      'languages': dict({
        'edges': list([
          dict({
            'node': dict({
              'alpha3Code': 'eng',
              'name': 'englanti',
              'translations': list([
                dict({
                  'languageCode': 'EN',
                  'name': 'English',
                }),
                dict({
                  'languageCode': 'FI',
                  'name': 'englanti',
                }),
                dict({
                  'languageCode': 'SV',
                  'name': 'Engelska',
                }),
              ]),
            }),
          }),
          dict({
            'node': dict({
              'alpha3Code': 'swe',
              'name': 'ruotsi',
              'translations': list([
                dict({
                  'languageCode': 'EN',
                  'name': 'Swedish',
                }),
                dict({
                  'languageCode': 'FI',
                  'name': 'ruotsi',
                }),
                dict({
                  'languageCode': 'SV',
                  'name': 'Svenska',
                }),
              ]),
            }),
          }),
          dict({
            'node': dict({
              'alpha3Code': 'fin',
              'name': 'suomi',
              'translations': list([
                dict({
                  'languageCode': 'EN',
                  'name': 'Finnish',
                }),
                dict({
                  'languageCode': 'FI',
                  'name': 'suomi',
                }),
                dict({
                  'languageCode': 'SV',
                  'name': 'Finska',
                }),
              ]),
            }),
          }),
        ]),
      }),
    }),
  })
# ---
